LFS-94: Data model

sdumitriu opened the pull request.

sdumitriu committed:
LFS-93 / LFS-94: Data model
First draft of the data model

sdumitriu committed:
LFS-93 / LFS-94: Data model
Fixed error in the default values for LONG properties
Without quotes, they are interpreted as DOUBLE values, which can't be assigned to a LONG property

sdumitriu committed:
LFS-93 / LFS-94: Data model
More data model work.

sdumitriu committed:
LFS-93 / LFS-94: Data model
A form's questionnaire reference is mandatory

sdumitriu committed:
LFS-93 / LFS-94: Data model
Questionnaires have a "title", not a "label"

sdumitriu committed:
LFS-93 / LFS-94: Data model
Added a "Textbox" display mode

sdumitriu committed:
LFS-93 / LFS-94: Data model
Most nodes must be referenceable

sdumitriu committed:
LFS-93 / LFS-94: Data model
The default dataType should be text

sdumitriu committed:
LFS-93 / LFS-94: Data model
Main entities need to be versionable

sdumitriu committed:
LFS-93 / LFS-94: Data model
Text cleanup

marta- commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -54,3 +54,456 @@
+
+  // The text of this option displayed to the user.
+  // This is not the value that gets stored when it is selected for an aswer, the "value" property is.
+  // Mandatory, every option needs a label.
```
Is it possible to default to `value` if `label` is missing, instead of making `label` mandatory?

sdumitriu replied:
Not in here, that's the responsibility either of the questionnaire builder code (for setting the label = value when defining the questionnaire), or of the form display code (for falling-back to the value if the label is missing).

sdumitriu replied:
Agreed that it doesn't have to be mandatory.

marta- commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -54,3 +54,456 @@
+  // Is answering this question mandatory?
+  // Even if this is set to false, setting a value to "minAnswers" can still make the question mandatory.
+  // Similarly, if "multiple" is true and "minAnswers" is 0, setting this to true doesn't have any effect.
+  - mandatory (boolean) = false autocreated
```
I don't like how other metadata (`multiple`, `minAnswers`) can cancel the effect of setting 'mandatory` to true.

I would only keep `minAnswers` and `maxAnswers` and infer `mandatory` and `multiple` from these:
`mandatory` = (`minAnswers` >= 1)
`multiple` = (`maxAnswers` != 1)

marta- commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -54,3 +54,456 @@
+  // - input: a simple input field where the user can type an answer
+  // - textbox: a large textbox for longer strings
+  // - list: a list of options that the user can choose from
+  // - list+input: a list of options, plus an "other" input where the user can add a non-standard answer
```
If I understand correctly, if a question is to have predefined options (and most will) we must
1/ set the display type to list[+input] AND
2/ provide the list of predefined options.

Seems a little redundant to me.

How about this:
* if the node does have children of type AnswerOptions, then automatically that is the list of options, no need to specify list in displayMode
* <need_to_find_a_good_name_for_this_property> : none|hidden|input|textbox with the definitions you provided above

sdumitriu committed:
LFS-93 / LFS-94: Data model
The label of an AnswerOption should be optional

sdumitriu committed:
LFS-93 / LFS-94: Data model
Removed "mandatory" and "multiple" fields from question, since they are redundant with "minAnswers" and "maxAnswers"
